auth/range_perm_cache.go

```diff
@@ -146,18 +87,44 @@ func checkKeyPerm(cachedPerms *unifiedRangePermissions, key, rangeEnd []byte, pe
 		plog.Panicf("unknown auth type: %v", permtyp)
 	}
 
-	requiredPerm := &rangePerm{begin: key, end: rangeEnd}
+	ivl := adt.NewStringAffineInterval(key, rangeEnd)
 
-	for _, perm := range tocheck {
-		if isSubset(requiredPerm, perm) {
-			return true
+	minBegin := adt.StringComparable("")
+	maxEnd := adt.StringComparable("\xff")
+
+	tocheck.Visit(ivl, func(n *adt.IntervalValue) bool {
+		if n.Ivl.Begin.Compare(minBegin) < 0 {
+			minBegin = n.Ivl.Begin.(adt.StringComparable)
+		}
+
+		if maxEnd.Compare(n.Ivl.End) < 0 {
+			maxEnd = n.Ivl.End.(adt.StringComparable)
 		}
+
```

Comment:
I think it's still possible to get something like `[a, d), [a, b), [c, f)`.
```go
isContiguous := true
var maxEnd, minBegin adt.Comparable
tocheck.Visit(ivl, func(n *adt.IntervalValue) bool { 
    if minBegin == nil { minBegin = n.Ivl.Begin; maxEnd = n.Ivl.End; return true}
    if maxEnd.Compare(n.ivl.Begin) < 0 { isContiguous = false; return false }
    if maxEnd.Compare(n.ivl.End) > 0 { maxEnd = n.ivl.End }
    return true
})
contains := isContiguous && maxEnd.Compare(ivl.End) >= 0 && minBegin.Compare(ivl.Begin) <= 0
```
?